src/p2p/communication.ts

```diff
@@ -0,0 +1,209 @@
+		this._node = await create(this._options);
+	}
+
+	async retry(): Promise<void> {
+		attempt(this, null, async () => {
+			for (const peer of this._retry) {
+				if (await this.connect(new Multiaddr(peer))) {
+					this._retry.delete(peer);
+				}
+			}
+		});
+	}
+
+	async start(): Promise<[Multiaddr, PeerId] | void> {
+		return attempt(this, undefined, async (node: Libp2p) => {
+			await node.start();
+
+			for (const peer of this._peers) {
+				if (!await this.connect(new Multiaddr(peer))) {
+					this._retry.add(peer);
+				}
+			}
+
+			this._node_addr = node.multiaddrs[0].toString();
+
+			return [
+				node.multiaddrs[0],
+				node.peerId,
+			];
+		})
+	}
+
+	async stop(): Promise<void> {
+		attempt(this, false, async (node: Libp2p) => {
+			try {
+				for (const peer of this._peers) {
+					node.hangUp(peer);
+				}
+
+				this.save_peers();
+
+				await node.stop();
+
+				return true;
+			} catch (error) {
+				logger.error(`Node failed to stop with error '${error}'.`);
+				return false;
+			}
+		});
+	}
+
+	async connect(ma: Multiaddr): Promise<boolean> {
+		return attempt(this, false, async (node: Libp2p) => {
+			try {
+				this._peers.add(ma.toString());
+				const conn = await node.dial(ma);
+				if (conn !== undefined) {
+					this._connections.add(conn);
+					return true;
+				}
+
+				return false;
+			} catch (error) {
+				logger.error(`Node failed to connect to ${ma} with error '${error}'.`);
+				return false;
+			}
+		});
+
+	}
+
+	async connect_from_details(ip: string, address: string, transport: string, port: string, peerID: string): Promise<boolean> {
+		return attempt(this, false, async (node: Libp2p) => {
+			const mas = `/${ip}/${address}/${transport}/${port}/p2p/${peerID}`;
+			const ma = new Multiaddr(`/${ip}/${address}/${transport}/${port}/p2p/${peerID}`);
+
+			try {
+				if (!this._peers.has(mas)) {
+					this._peers.add(mas);
+				}
+
+				const conn = await node.dial(ma);
+				if (conn !== undefined) {
+					this._connections.add(conn);
+					return true;
+				}
+
+				return false;
+			} catch (error) {
+				this._peers.delete(mas);
+				logger.error(`Node failed to connect to ${ma} with error '${error}'.`);
+				return false;
+			}
+		});
+	}
+
+	async handle_incoming(protocol: string, callback: (source: AsyncIterable<Uint8Array | BufferList>) => Promise<void>): Promise<void> {
+		await attempt(this, null, async (node: Libp2p) => {
+			node.handle(protocol, async ({ stream }) => {
+				await callback(stream.source);
+			});
+		});
+	}
+
+	async send(protocol: string, data: Uint8Array[]): Promise<void> {
+		attempt(this, null, async () => {
+			for (const connection of this._connections) {
+				const { stream } = await connection.newStream(protocol);
+				await stream.sink(createAsyncIterable(data));
+			}
+		});
+	}
+
+	save_peers(): void {
+		let peers = Array.from(this._peers).sort();
+
+		const json = JSON.stringify(peers, null, 4);
+		try {
+			writeFileSync(this._peers_file, json);
```

Comment:
Would probably be better to use the async counterpart (there is a promisfied version of writeFile) in order to not block the thread